src/persist-client/src/read.rs

```diff
@@ -0,0 +1,149 @@
+    /// read at times not greater or equal to `new_since`.
+    ///
+    /// This may trigger (asynchronous) compaction and consolidation in the
+    /// system. A `new_upper` of the empty antichain "finishes" this collection,
+    /// promising that no more data will ever be read by this handle.
+    //
+    // WIP AntichainRef?
+    pub fn downgrade_since(&self, new_since: Antichain<T>) {
+        todo!("{:?}", new_since)
+    }
+
+    /// Returns an ongoing stream of updates to a collection.
+    ///
+    /// The subscription includes all data at times greater than `as_of`.
+    /// Combined with [Self::snapshot] it will produce exactly correct results:
+    /// the snapshot is the TVCs contents at `as_of` and all subsequent updates
+    /// occur at exactly their indicated time. The recipient should only
+    /// downgrade their read capability when they are certain they have all data
+    /// through the frontier they would downgrade to.
+    //
+    // WIP: This is an opportunity to push down Projection information to save
+    // bandwidth. My thinking so far is that Map and Filter should stay
+    // STORAGE-level concerns.
+    //
+    // WIP: My first attempt here was to return `impl Stream` but that didn't
+    // compile with `todo!`. I don't yet have an opinion on which is better.
+    pub async fn listen(as_of: Antichain<T>) -> Result<Listen<K, V, T, D>, Error> {
+        todo!("{:?}", as_of);
+    }
+
+    /// Returns a snapshot of the collection at `as_of`.
+    ///
+    /// This command returns the contents of this collection as of `as_of` once
+    /// they are known. This may be in the future if `as_of` is greater or equal
+    /// to the current `upper` of the collection. The recipient should only
+    /// downgrade their read capability when they are certain they have all data
+    /// through the frontier they would downgrade to.
+    ///
+    /// This snapshot may be split into a number of partitions, each of which
+    /// may be exchanged (including over the network) to load balance the
+    /// processing of this snapshot. These partitions are usable by anyone with
+    /// access to the collection's [crate::Location]. The `len()` of the
+    /// returned `Vec` is `num_parts`.
+    //
+    // WIP This is an opportunity to push down Projection information to save
+    // bandwidth. My thinking so far is that Map and Filter should stay
+    // STORAGE-level concerns.
+    pub async fn snapshot(
+        &self,
+        as_of: Antichain<T>,
+        num_parts: NonZeroUsize,
+    ) -> Result<Vec<SnapshotPart>, Error> {
+        todo!("{:?}{:?}", as_of, num_parts);
+    }
+
+    /// Trade in an exchange-able [SnapshotPart] for its respective data.
+    //
+    // WIP: Should this live on ReadHandle or somewhere else? E.g. I think it
+    // could also live on Client.
+    //
+    // WIP: Figure out the error story here. Retries should move inside this, but
```

Comment:
Retry-er must be observable. Petros's initial reaction is to expose an error on whether it's retryable